### ISSUE TITLE
RAM usage keeps increasing when using `loss.detach().cpu().numpy()`

### ISSUE THREAD

**Issue by mspinaci** [googlecolab/colabtools#4150]
**Describe the current behavior**
I created a small colab with a straightforward PyTorch training loop of a tiny model (Logistic regression). If, during training, I log the loss by doing:
```
train_loss.append(loss_result.detach().cpu().numpy())
```
then the RAM usage blows up very quickly:
<img width="1697" alt="Screenshot 2023-11-13 at 22 35 38" src="https://github.com/googlecolab/colabtools/assets/32176505/7c5e5106-eeca-4fa7-a93a-bb3e7e26d79f">
(GPU memory is barely used at all because the model is tiny)
The issue is fixed if I replace that line by:
```
train_loss.append(loss_result.item())
```
or
```
train_loss.append(float(loss_result.detach().cpu().numpy()))
```
the issue goes away and memory gets released as expected.

The only reason I can imagine is that also the whole model gradients get somehow copied to CPU and not released? It doesn't make too much sense though (especially since there's `.detach()` being called before `.cpu()`)

**Describe the expected behavior**
I would expect to be able to log numpy array without the memory blowing up

**What web browser you are using**
Chrome

**Additional context**
Here is the colab:
https://colab.research.google.com/drive/1dfZtbC-2AjQkK1jt5awrClW_UZRAKBQv?usp=sharing

**Comment by cperry-goog** [comment]
This does not appear to be a Colab-specific issue; I can't imagine a world where this behaves differently on another machine. Can you confirm this only occurs on Colab, and not any other VM with the same set of dependencies installed as Colab?

**Comment by mspinaci** [comment]
I'm sorry, I can't test the same setup elsewhere, I'm unsure how I could set up such a virtual machine...

But while I agree that this would be weird for a Colab-related error, I keep getting out-of-RAM errors for seemingly unrelated issues, and I've never seen any such errors on any other platform (where I've used similar libraries). The latest one I've found triggered with just the following 5 lines of code, run from scratch on a Colab instance (CPU or GPU runtime, doesn't matter):

```
import torch
from torchvision.models import ResNet34_Weights
preprocess = ResNet34_Weights.IMAGENET1K_V1.transforms()
x = torch.rand(10, 160, 160, 3)
preprocess(x)
```

While this might in theory be a torch / torchvision issue, these are quite common libraries, and I have never heard of such errors anywhere else than Colab...
<img width="1207" alt="Screenshot 2023-11-26 at 23 57 15" src="https://github.com/googlecolab/colabtools/assets/32176505/79d19939-5215-4f94-8e27-64ed763b4c42">
